src/hotspot/share/cds/dumpTimeClassInfo.cpp

```diff
@@ -152,11 +152,17 @@ DumpTimeClassInfo* DumpTimeSharedClassTable::allocate_info(InstanceKlass* k) {
 }
 
 DumpTimeClassInfo* DumpTimeSharedClassTable::get_info(InstanceKlass* k) {
-  assert(!k->is_shared(), "Do not call with shared classes");
+  if (DumpSharedSpaces) {
+    assert(!k->is_shared(), "Do not call with shared classes");
+  } else {
+    assert(DynamicDumpSharedSpaces, "sanity");
+  }
   DumpTimeClassInfo* p = get(k);
-  assert(p != nullptr, "we must not see any non-shared InstanceKlass* that's "
-         "not stored with SystemDictionaryShared::init_dumptime_info");
-  assert(p->_klass == k, "Sanity");
+  if (DumpSharedSpaces) {
+    assert(p != nullptr, "we must not see any non-shared InstanceKlass* that's "
+           "not stored with SystemDictionaryShared::init_dumptime_info");
+    assert(p->_klass == k, "Sanity");
+  }
```

Comment:
We have lots of code that looks like this:

```
DumpTimeClassInfo* info = get_info(k);
info->_clsfile_size  = cfs->length();
```

The assumption is that `k` must have a `DumpTimeClassInfo`, which means that `k` was created by the `ClassFileParser` (and not loaded with `SystemDictionary::load_shared_class()`).

With the above change, `get_info()` could potentially return null, which means that the caller now need to check and handle for the null return. This will make the code more complicated.

I think a better fix would be to avoid calling `SystemDictionaryShared::set_class_has_failed_verification()` on classes that are shared. Maybe the condition in `MetaspaceShared::try_link_class()` can be changed like this?

```
- if (ik->is_loaded() && !ik->is_linked() && ik->can_be_verified_at_dumptime() &&
+ if (!ik->is_shared() && ik->is_loaded() && !ik->is_linked() && ik->can_be_verified_at_dumptime() &&
```

Reply:
With the above suggestions, it means that we don't try to link any shared classes which had failed linking before.
It seems I could revert the other changes too.